Add delete_in_batches for batch deletions

shayonj opened the pull request:
### Motivation / Background

Currently, the default option for `order` is `asc` and can also be set to `desc` or passed an array with those values. This works great in most cases, however sometimes forcing the `ORDER` clause for the primary key can result into a sort + heap scan in Postgresql, by passing an index lookup and may cause a performance tax and in those situations where we don't care about order, ~it'd be lovely to continue using `in_batches` with no order instead of having to write custom SQL.~

**Update:** Based on feedback from @jeremy, @p8, and @skipkayhil, this PR now introduces `delete_in_batches` instead of adding `order: nil` support.

### Solution

Add `delete_in_batches` method:

```ruby
Event.where(story_run_id: ids).delete_in_batches
Event.where(processed: true).delete_in_batches(of: 5000)
```

For bulk deletes where order doesn't matter, something like can actually be faster because the lookups just hit the primary key index.

Added context (from the conversations below): This is cleaner than `order: nil` because the intent is explicit (as this is for deletions) and we don't need to modify the order function either. 

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

jeremy reviewed (commented):
Really clever usage with batched deletions.

I'd document it separately as a special-case advanced usage.

Can you `.reorder(nil)` to hack out the ordering? If there's an escape hatch that works, it's probably worth not muddying the method.

shayonj committed:
Add delete_in_batches for batch deletions

Adds a method to delete records in batches without cursor tracking
or ORDER BY. Since deleted records are excluded from subsequent
batches, no ordering is required.

  Event.where(processed: true).delete_in_batches(of: 1000)

Returns the total number of rows deleted.

shayonj commented:
👋🏾 Hello team - wanted to see if this proposal was still something you were interested in? Happy to rebase accordingly.

fxn commented:
When working with batches, I am always concerned about the unlikely but possible race conditions from rows being updated or inserted in busy tables.

Would you consider breaking the loop when the number of deleted rows is strictly less than the batch size? Race conditions still possible, but the cut is clear and there is one less statement.

**EDIT:** Nah, forget it, the query is arbitrary and could return duplicates.
